iothub/device/tests/Authentication/SharedAccessSignatureBuilderTest.cs

```diff
@@ -55,5 +55,131 @@ public void TestBuildSasToken_ShouldReturnSasToken()
+            SharedAccessSignature _ = SharedAccessSignatureParser.Parse(sasTokenString);
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(FormatException))]
+        public void TestBuildSasToken_ShouldReturnSasToken_MalformedKeyword()
+        {
+            string audience = "iothub.test/devices/device1/modules/module1";
+            string signature = "signature";
+
+            DateTime expiresOn = DateTime.UtcNow.AddMinutes(10);
+            TimeSpan secondsFromBaseTime = expiresOn.Subtract(SharedAccessSignatureConstants.EpochTime);
+            string expiry = secondsFromBaseTime.TotalSeconds.ToString(CultureInfo.InvariantCulture);
+
+            string sasTokenString = SharedAccessSignatureBuilder.BuildSignature(null, null, null, TimeSpan.Zero, audience, signature, expiry);
+            sasTokenString = sasTokenString.Replace(SharedAccessSignatureConstants.SharedAccessSignature, SharedAccessSignatureConstants.SharedAccessSignature + "_");
+            SharedAccessSignature _ = SharedAccessSignatureParser.Parse(sasTokenString);
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(FormatException))]
+        public void TestBuildSasToken_ShouldReturnSasToken_MalformedFields()
+        {
+            string audience = "iothub.test/devices/device1/modules/module1";
+            string signature = "signature";
+
+            DateTime expiresOn = DateTime.UtcNow.AddMinutes(10);
+            TimeSpan secondsFromBaseTime = expiresOn.Subtract(SharedAccessSignatureConstants.EpochTime);
+            string expiry = secondsFromBaseTime.TotalSeconds.ToString(CultureInfo.InvariantCulture);
+
+            string sasTokenString = SharedAccessSignatureBuilder.BuildSignature(null, null, null, TimeSpan.Zero, audience, signature, expiry);
+            sasTokenString = sasTokenString.Replace(SharedAccessSignatureConstants.AudienceFieldName + SharedAccessSignatureConstants.KeyValueSeparator, "");
+            SharedAccessSignature _ = SharedAccessSignatureParser.Parse(sasTokenString);
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(IotHubClientException))]
+        public void TestBuildSasToken_ShouldReturnSasToken_expired()
+        {
+            string audience = "iothub.test/devices/device1/modules/module1";
+            string signature = "signature";
+
+            DateTime expiresOn = DateTime.UtcNow.AddMinutes(10);
+            TimeSpan secondsFromBaseTime = expiresOn.Subtract(SharedAccessSignatureConstants.EpochTime);
+            string expiry = TimeSpan.FromSeconds(-1).TotalSeconds.ToString(CultureInfo.InvariantCulture);
+
+            string sasTokenString = SharedAccessSignatureBuilder.BuildSignature(null, null, null, TimeSpan.Zero, audience, signature, expiry);
+
+            SharedAccessSignature token = SharedAccessSignatureParser.Parse(sasTokenString);
+
+            Assert.AreEqual(WebUtility.UrlDecode(audience), token.Audience);
```

Comment:
Please use FluentAssertions instead.

```suggestion
            token.Audience.Should().Be(WebUtility.UrlDecode(audience));
```